File: 0875-koko-eating-bananas/0875-koko-eating-bananas.java
class Solution {
    public int calculateTime(int speed, int[] piles){
        int t = 0;
        for(int num : piles){
            t += Math.ceil((double)(num) /(double)(speed));
        }
        return t;
    }
    
    public int minEatingSpeed(int[] piles, int h) {
        int maxPile = 0;
        for(int p : piles){
            maxPile = Math.max(maxPile, p);
        }
        int start = 1;
        int end = maxPile;
        while(start <= end){
            int mid = start + ((end - start) >> 1);
            int t = calculateTime(mid, piles);
            if(t <= h){
                end = mid - 1;
            } else{
                start = mid + 1;
            }
        }
        return start;
    }
}

File: 0875-koko-eating-bananas/0875-koko-eating-bananas.py
class Solution:
    def calculate_time(self, speed, piles):
        t = 0
        for num in piles:
            t += -(-num // speed)  # Equivalent to Math.ceil((double)(num) /(double)(speed))
        return t
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        max_pile = max(piles)
        start, end = 1, max_pile

        while start <= end:
            mid = start + (end - start) // 2
            t = self.calculate_time(mid, piles)

            if t <= h:
                end = mid - 1
            else:
                start = mid + 1

        return start
     